proposals/5/README.md

```diff
@@ -0,0 +1,103 @@
+
+At the moment, the source code that comprises the main JX binary is tightly coupled and difficult to work on.  Changes have
+unintended consequences and this makes the codebase fragile.
+
+One of the design goals in Accelerate is to have a **Loosely Coupled Architecture** - in particular: 
+
+_The goal is for your architecture to support the ability of teams to get their work done—from design through to deployment—without requiring high-bandwidth communication between teams._
+
+## 1.2 Background
+
+The main [JX repository](https://github.com/jenkins-x/jx) contains over 250k lines of code, creating Apps / Extensions / Binary Plugins is difficult without adding a dependency back onto the JX repo.
+
+It should be possible to separate the repository out into a series of reusable libraries so only what is needed can be imported.
+
+## 2. Design
+
+There are currently a number of proposed repositories/modules, some of these are currently in flight.
+
+### Existing Modules
+
+* [jx-api](https://github.com/jenkins-x/jx-api): JX CRDs / Installation Requirements
+* [jx-kube-client](https://github.com/jenkins-x/jx-kube-client): A helper module to create a Kubernetes rest config
+* [jx-vault-client](https://github.com/jenkins-x/jx-vault-client): longer term, it may make sense to deprecate this.
+* [go-scm](https://github.com/jenkins-x/go-scm): All 
+
+### Proposed Modules
+
+* jx-kube: It may make sense to refactor some of the higher level utils/kube functions within JX into its own package to make them reusable.
+* Version Streams: Would these make sense to be moved out of JX
+* Gitter?
+* Updatebot (jx create pr)? This would have value outside of the core JX codebase 
+* Storage APIs?
+* Cloud APIs?
+
+## 2.1. Action Plan
+
+1. *Agree on a way forward!*
+1. Ensure [Kubernetes Dependency Updates PR](https://github.com/jenkins-x/jx/pull/7313) is merged.
+1. Re-introduce [jx-api](https://github.com/jenkins-x/jx-api) as a dependency within [jx](https://github.com/jenkins-x/jx)
+1. Test and Release
+
+Once we have completed this once, we should be able to repeat the following process a number of times
+
+1. Identify code to be extracted
+1. Extract code into new repo
+1. Release new repository
+1. Reintroduce library as a dependency inside JX
+1. Repeat as required.
+
+## 2.2 Considerations for New Repositories
+
+All new repositories should aim for the following:
+
+* Basic documentation to explain how the module should be used
+* A full set of [linters](https://golangci-lint.run/usage/linters), suggesting (asciicheck bodyclose deadcode dogsled dupl errcheck goconst gofmt goimports gosec gosimple govet ineffassign interfacer misspell staticcheck structcheck typecheck unconvert unparam unused varcheck), ideally add gocyclo, nestif & gocritic
+* Tests (we should define a base level for code coverage, I'd expect this to be quite low initially)
+* [Go Report Card](https://goreportcard.com/)
+* Should aim to use [Semantic Versioning](https://semver.org/)
```

Comment:
How are we going to do this given that we cannot really do it with current Jenkins X? Are we doing manual releases for now? 

Would we still apply the every single merge to master creates a release approach?

Reply:
My hope, is that using Jenkins-X internally for a real use case (proper dogfooding), we'll realise we have a gap when it comes to SemVer and can implement whats needed.

Reply:
IMO we know we have a gap. I would go the other way around and provide these new libraries with a good start, doing from the beginning what we want to do in the long run.
Given that we are talking about Go libraries an actual release is not much more than creating and pushing a tag after something successfully merged to master. This can easily be done via a script which is executed on demand. 
Then there would be the same motivation to close the gap to fully integrate this into Jenkins X and the benefit that you do things right from the beginning.
Also imo it would be beneficial to not releasing on every single push to master (another feature gap imo).